### PR TITLE
Fixed #29750 -- Added View.setup() hook.

### PR DESCRIPTION
**Pull request by francoisfreitag**
Allows encapsulating logic in View mixins for reuse in child classes.

The standard request processing cycle is:

1. Concrete view
2. Mixins
3. Generic view

Before `init()`, it was not possible to reuse self attributes set in parent views and mixins from the concrete view, because calling the parent method generates a `HttpResponse`.

https://code.djangoproject.com/ticket/29750

### CODE REVIEW

File: django/views/generic/base.py [hunk 1]
```diff
@@ -62,9 +62,7 @@ def view(request, *args, **kwargs):
             self = cls(**initkwargs)
             if hasattr(self, 'get') and not hasattr(self, 'head'):
                 self.head = self.get
-            self.request = request
-            self.args = args
-            self.kwargs = kwargs
+            self.init(request, *args, **kwargs)
```
Comment (carltongibson) [inline django/views/generic/base.py]: Since we moved setting these attributes into `init()`, is it worth an `assert` afterwards to make sure I didn't override it and forget to call `super()`?

Comment (francoisfreitag) [reply django/views/generic/base.py]: Sure, it avoids inattention mistakes and diving in Django code to understand what's wrong. Without the assert, the error is very explicit (e.g. `AttributeError: 'LoginView' object has no attribute 'request'`), but an assert message helps understanding the cause.

If we add the assertion, do you think the note in the doc about calling `super()` is still helpful?

Comment (carltongibson) [reply django/views/generic/base.py]: Personally yes. (But maybe the examples in the docs will be enough)

**Comment by francoisfreitag** [comment]
Hi Carlton,

Thank you very much for reviewing the PR and your input.

> I was surprised to see you still override `dispatch()` in `ImpersonationWithMessageMixin`.

In my mind, `init()` is responsible for performing view attribute initialization and nothing else. The idea behind using `dispatch()` in the example is to avoid encouraging developers to perform other tasks in `init()`.

> [...] the idea here was to allow us to leave dispatch alone?

Not exactly, the idea is to provide parent views and mixins with a way to abstract logic from child classes, but let children control the response that's generated.
In the example, the child `ImpersonationWithMessageMixin` must know the user being impersonated, that is set by parent `ImpersonationMixin`. To execute the parent logic that sets `self.user` without `init()`, the child needs to call `super().dispatch()`, which generates a `HttpResponse`.

---
Maybe adding a message is not the best example, the example could be updated to check permissions. For example, we could return a `HttpResponseForbidden` when `not self.user.is_active`. That way, an admin can't record interest for inactive users.
Checking access permissions could be a better use of dispatch, and it is simpler (doesn't involve the message framework). I'll update the example before to add it to the documentation.

---
In the ticket, I had hopes to move all `self.object = get_object()` to `init()`, to avoid duplication in view methods, such as `BaseUpdateView`'s `get()` and `post()`. However, this is backward incompatible: `init()` would trigger `get_object ()` earlier in the view lifecycle and may issue DB queries when none were necessary before, because `dispatch()` may raise a `PermissionDenied`, then the `get()` or `post()` method wouldn't execute `get_object()`.

**Comment by carltongibson** [comment]
> I had hopes to move all self.object = get_object() to init(), to avoid duplication in view methods, such as BaseUpdateView's get() and post().

I'd be super -1 on this. The `get_object()` call is an integral part of the view logic. Moving to `init()` just obscures the flow. A bit of repetition actually makes for clearer code here. (And in many places TBH. It's quite easy to take DRY too far. IMO)

File: docs/topics/class-based-views/mixins.txt [hunk 2]
```diff
@@ -686,3 +686,123 @@ Because of the way that Python resolves method overloading, the call to
...
+            if not self.user.is_active:
+                return HttpResponseForbidden('User is inactive.')
+            # Generates a HttpResponse, must be called after user status check.
+            return super().dispatch(request, *args, **kwargs)
+
+Using :meth:`~django.views.generic.base.View.init()`
+----------------------------------------------------
+
+The :meth:`~django.views.generic.base.View.init()` hook is designed to run
+authorization checks before to prepare a :class:`~django.http.HttpResponse`.
```
Comment (carltongibson) [inline docs/topics/class-based-views/mixins.txt]: `before dispatch`?

File: docs/topics/class-based-views/mixins.txt [hunk 3]
```diff
@@ -686,3 +686,123 @@ Because of the way that Python resolves method overloading, the call to
...
+.. code-block:: python
+    :caption: views.py
+
+    class ActiveUserImpersonationMixin(ImpersonationMixin):
+        def dispatch(self, request, *args, **kwargs):
+            # CAUTION: Raises AttributeError, self.user is not (yet) set.
+            if not self.user.is_active:
+                return HttpResponseForbidden('User is inactive.')
+            # Generates a HttpResponse, must be called after user status check.
+            return super().dispatch(request, *args, **kwargs)
```
Comment (carltongibson) [inline docs/topics/class-based-views/mixins.txt]: Hmmm. I still have real concerns about showing users that they should move core view logic into dispatch... It might work in a particular shop, but I worry we're just teaching people to put their logic all over the place, rather than straight-line in front of them. 

(Just an observation.)

**Comment by jdufresne** [comment]
> What do others think? (I know you like it @jdufresne)

Yeah, I do like it and I think François has done a great job here. I'm +1 on merging this once the docs meet your approval.

**Comment by charettes** [comment]
> What do others think?

My only concern here is the `init()` method name which doesn't make it clear it isn't directly related to `__init__()` IMO. Were alternative names such as `setup()` considered?

**Comment by carltongibson** [comment]
My concern isn’t really the feature at this point. I wouldn’t go this way, but it’s fine and I can see the utility if it’s the style you want to use. So 👍 there. 

I am concerned about the docs example: I think a lot of users take the docs as being normative, as saying _do this!_; I think unless you really know what you’re doing putting parts of your view logic in dispatch is not something you should do. (You should just keep it in your view function, even if that means a bit of repetition calling `stepOne()`, `stepTwo()` and so on.) Thus I think the example is tacitly providing the wrong advice for most people. (A small docs entry highlighting the feature for someone who’s already looking for it would be better I think.)

If no one else is feeling this then I’m happy to let it pass. (It’s quite possible I’m just too old. 🙂) But that’s why I asked for other opinions (because I can’t shake the feeling, having given it time enough for a change of heart.) I’d basically like to recuse myself here. Thanks for your help all.

**Comment by francoisfreitag** [comment]
As for the name, I kept `init()` because it's short and the hook's purpose is to `init`ializes view attributes. I am happy to change it to `setup` if it helps understand what this hook can be used for. It's clear that `setup()` will run before `dispatch()` and I would not expect `setup()` to return anything.

`init()` was the name suggested in the previous mailing list discussion. For a while, I had `pre_dispatch()` in mind, but that's not entirely what this hook does. It allows parent code to execute before child code, so I dropped that idea. That's the names that made sense to me, I'm definitely open to alternatives.

**Comment by carltongibson** [comment]
Hi @francoisfreitag Thanks for the follow-up. 

I don't know that we need to go to the mailing list. Jon like feature, Simon has expressed support, I see the use of it (even if it's not my thing per se) — let's go with it. 

Only thing now is the docs. (And maybe a rename.) I like the changes you've made: they're much more in line with how I'd present. 

Allow me to give it another once-over and we'll see about getting it in. 

Thanks for effort!

File: django/views/generic/base.py [hunk 4]
```diff
@@ -81,7 +82,7 @@ def view(request, *args, **kwargs):
         return view
 
     def init(self, request, *args, **kwargs):
-        """Initialize self attributes shared by all view methods."""
+        """Initialize key attributes shared by all view methods."""
```
[the remark this answers is not in the document]
Comment (carltongibson) [reply django/views/generic/base.py]: Yep. Fine. (If you're editing can you squash?)

**Comment by carltongibson** [comment]
My thought was that it's fine as-is, i.e. minus the example. (I think people looking for it can come up with their own use-cases.) There's plenty of discussion on the ticket (I think). 

I also thought that your use-case **would** make a good blog-post, on your own or company blog.

**Comment by carltongibson** [comment]
> Were alternative names such as setup() considered?

So I was just about to mark this RFC when I remembered this. _Naming things..._

File: django/views/generic/base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (francoisfreitag) [reply django/views/generic/base.py]: For sure, updated.

File: django/views/generic/base.py [hunk 5]
```diff
@@ -62,9 +62,13 @@ def view(request, *args, **kwargs):
             self = cls(**initkwargs)
             if hasattr(self, 'get') and not hasattr(self, 'head'):
                 self.head = self.get
-            self.request = request
-            self.args = args
-            self.kwargs = kwargs
+            self.init(request, *args, **kwargs)
+            assert hasattr(self, 'request'), (
```
Comment (timgraham) [inline django/views/generic/base.py]: I think `if not hasattr(...)` and raising an exception would be more in line with Django's style.

File: django/views/generic/base.py [hunk 6]
```diff
@@ -62,9 +62,13 @@ def view(request, *args, **kwargs):
...
+                "%s instance has no 'request' attribute. Did you override "
+                "'init()' and forget to call 'super()'?" % (
+                    cls.__name__,
```
Comment (timgraham) [inline django/views/generic/base.py]: You can omit the tuple and just put this on the previous line.

File: docs/ref/class-based-views/base.txt [hunk 7]
```diff
@@ -70,11 +71,17 @@ MRO is an acronym for Method Resolution Order.
...
+        the view's ``request`` attribute, and any positional and/or keyword
+        arguments :ref:`captured from the URL pattern
+        <how-django-processes-a-request>` to the ``args`` and ``kwargs``
+        attributes, respectively. Then :meth:`dispatch` is called.
+
+    .. method:: init(request, *args, **kwargs)
+
+        Initializes view instance attributes: ``self.request``, ``self.args``
+        and ``self.kwargs`` prior to :meth:`dispatch`. Child classes overriding
+        this method, must call ``super()``.
```
Comment (timgraham) [inline docs/ref/class-based-views/base.txt]: chop comma

File: docs/ref/class-based-views/base.txt [hunk 8]
```diff
@@ -70,11 +71,17 @@ MRO is an acronym for Method Resolution Order.
...
-        is called.
+        :meth:`init` method assigns the :class:`~django.http.HttpRequest` to
+        the view's ``request`` attribute, and any positional and/or keyword
+        arguments :ref:`captured from the URL pattern
+        <how-django-processes-a-request>` to the ``args`` and ``kwargs``
+        attributes, respectively. Then :meth:`dispatch` is called.
+
+    .. method:: init(request, *args, **kwargs)
+
+        Initializes view instance attributes: ``self.request``, ``self.args``
```
Comment (timgraham) [inline docs/ref/class-based-views/base.txt]: a, b, and c (missing comma after 'b')

also missing `.. versionadded:: 2.2` annotation

Also you mentioned a use case in the release notes but not here.

File: docs/ref/class-based-views/generic-display.txt [hunk 9]
```diff
@@ -25,16 +25,17 @@ many projects they are typically the most commonly used views.
...
-    2. :meth:`~django.views.generic.base.View.http_method_not_allowed()`
-    3. :meth:`~django.views.generic.base.TemplateResponseMixin.get_template_names()`
-    4. :meth:`~django.views.generic.detail.SingleObjectMixin.get_slug_field()`
-    5. :meth:`~django.views.generic.detail.SingleObjectMixin.get_queryset()`
-    6. :meth:`~django.views.generic.detail.SingleObjectMixin.get_object()`
-    7. :meth:`~django.views.generic.detail.SingleObjectMixin.get_context_object_name()`
-    8. :meth:`~django.views.generic.detail.SingleObjectMixin.get_context_data()`
-    9. ``get()``
-    10. :meth:`~django.views.generic.base.TemplateResponseMixin.render_to_response()`
+    1. :meth:`~django.views.generic.base.View.init()`
```
Comment (timgraham) [inline docs/ref/class-based-views/generic-display.txt]: Please submit a separate PR that we can merge before this one that replaces the numbers with `#.`. That'll clean this up.

File: docs/releases/2.2.txt [hunk 10]
```diff
@@ -165,7 +165,10 @@ Forms
 Generic Views
 ~~~~~~~~~~~~~
 
-* ...
+* A new :meth:`~django.views.generic.base.View.init` hook initializes
```
Comment (timgraham) [inline docs/releases/2.2.txt]: I think "The new" is more inline with other notes.

`View.init <django.views.generic.base.View.init>` would render "View" also which would help.

File: docs/releases/2.2.txt [hunk 11]
```diff
@@ -165,7 +165,10 @@ Forms
...
+  :class:`~django.views.generic.base.View` attributes before calling
```
Comment (timgraham) [inline docs/releases/2.2.txt]: I would replace "class..." with "view" (no link)

File: docs/releases/2.2.txt [hunk 12]
```diff
@@ -165,7 +165,10 @@ Forms
...
+  :meth:`~django.views.generic.base.View.dispatch`. Allows mixins to setup
```
Comment (timgraham) [inline docs/releases/2.2.txt]: It allows....

File: tests/generic_views/test_base.py [hunk 13]
```diff
@@ -223,6 +235,22 @@ def test_args_kwargs_request_on_self(self):
...
+    def test_overridden_init(self):
+        response = CheckInitView.as_view()(self.rf.get('/'))
+        self.assertEqual(response.status_code, 200)
+
+    def test_not_calling_parent_init_raises(self):
+        class TestView(View):
+            def init(self, request, *args, **kwargs):
+                pass  # Not calling parent init()
+
+        with self.assertRaises(AssertionError) as cm:
```
Comment (timgraham) [inline tests/generic_views/test_base.py]: Use `self.assertRaisesMessage`.

File: tests/generic_views/test_base.py [hunk 14]
```diff
@@ -44,6 +44,18 @@ def dispatch(self, request, *args, **kwargs):
         return super().dispatch(request, *args, **kwargs)
 
 
+class SetAttributeMixin:
```
Comment (timgraham) [inline tests/generic_views/test_base.py]: I would put this in the one test they're used in unless you think they have potential to be used in future tests.

**Comment by carltongibson** [comment]
Hey Tim. Thanks for the review. Do you have any thoughts on the naming question?

**Comment by timgraham** [comment]
No strong opinion. One idea is that it the name `init` could create some ambiguity with `__init__` when speaking.

File: docs/ref/class-based-views/generic-display.txt (hunk 9 above)
Comment (francoisfreitag) [reply docs/ref/class-based-views/generic-display.txt]: https://github.com/django/django/pull/10649

**Comment by carltongibson** [comment]
> No strong opinion...

OK, let's call that a +0.5 🙂 I'll say the same for Simon and myself (for total +1.5)

@francoisfreitag Could you change for Simon's suggestion of `setup()`. I don't suppose there's much in it but it avoids the verbal confusion as Tim says? 

Then if you could address Tim's review comments I think we're there. 🎉

**Comment by francoisfreitag** [comment]
I don't mind changing to `setup()`. Thanks for the feedback, I'll update the PR in a few days.

**Comment by francoisfreitag** [comment]
Early commit to address [comment "Use `self.assertRaisesMessage`."](https://github.com/django/django/pull/10427/#discussion_r233664604) introduced as https://github.com/django/django/pull/10714.

Ready for review.

**Comment by francoisfreitag** [comment]
Sorry @timgraham, I push-forced your edits. I restored them from https://github.com/django/django/compare/53d05c3d166eb2f5e410b7a6c2c947a6480b1c50..c1022f47e051924288d2d5cfbdf7c7cfecd59a79, but would not mind a second set of eyes.

I'll make sure to use `--force-with-lease` in the future.

**Comment by francoisfreitag** [comment]
The commit message was reverted. I was only able to restore the first line.

**Comment by francoisfreitag** [comment]
Note: I did make a change on my side (`CheckInitView` -> `CheckSetupView`).

File: docs/ref/class-based-views/base.txt [hunk 15]
```diff
@@ -70,11 +71,21 @@ MRO is an acronym for Method Resolution Order.
...
+
+    .. method:: setup(request, *args, **kwargs)
+
+        .. versionadded:: 2.2
+
+        Initializes view instance attributes: ``self.request``, ``self.args``,
+        and ``self.kwargs`` prior to :meth:`dispatch`.
+
+        Overriding this method allows mixins to setup instance attributes for
+        reuse in child classes. Child classes must call ``super()``.
```
Comment (carltongibson) [inline docs/ref/class-based-views/base.txt]: I preferred the shorter version we had earlier: 

```
Initializes view instance attributes: ``self.request``, ``self.args``
and ``self.kwargs`` prior to :meth:`dispatch`. Child classes overriding
this method must call ``super()``.
```

Comment (francoisfreitag) [reply docs/ref/class-based-views/base.txt]: I restored it from https://github.com/django/django/compare/53d05c3d166eb2f5e410b7a6c2c947a6480b1c50..c1022f47e051924288d2d5cfbdf7c7cfecd59a79#diff-2df22e647d63040c7db3eca921c888fdL87, I believe this change was part of @timgraham's cosmetic edits.

The longer version explains the purpose of this hook, that's why I have a slight preference for it.

File: tests/generic_views/test_base.py [hunk 16]
```diff
@@ -233,6 +233,32 @@ def test_args_kwargs_request_on_self(self):
...
+                assert hasattr(self, 'attr')
+                return super().dispatch(request, *args, **kwargs)
+
+        response = CheckSetupView.as_view()(self.rf.get('/'))
+        self.assertEqual(response.status_code, 200)
+
+    def test_not_calling_parent_setup_error(self):
+        class TestView(View):
+            def setup(self, request, *args, **kwargs):
+                pass  # Not calling supre().setup()
```
Comment (todd-dembrey) [inline tests/generic_views/test_base.py:253]: `s/supre/super/`

Comment (timgraham) [reply tests/generic_views/test_base.py:253]: (was fixed in b916c27f9a83c019ff85132712267508f126ca83)
